File: lua/documentos/valida_cnpj.lua
#!/usr/bin/env lua

--local utils, calculos;

if type(calculos) == "nil" then
    --calculos = require "calculos";
    local calculos = require "calculos";
    if type(calculos) == "nil" then
        print("Falha na carga do módulo calculos");
        return utils.STATE_FAIL;
    end
end

if type(utils) == "nil" then
    --utils = require "utils";
    local utils = require "utils";
    if type(utils) == "nil" then
        print("Falha na carga do módulo utils");
        return utils.STATE_FAIL;
    end
end

--print(type(calculos)..type(utils));

____VALIDA_CNPJ____ = "loaded";

local CNPJ_TAMANHO = 14;
local ERRO_CARACTER = 1;
local ERRO_TAMANHO=  2;
local ERRO_REPETICAO = 3;
local ERRO_DIGITO = 4;
local ERRO_MSG_SAIDA = "";

valida_cnpj = {};

valida_cnpj.CODIGO_SAIDA = 0;

function valida_cnpj.print_usage()
    local __usage = [[

    Sintaxe: 
        valida_cnpj(<cnpj>);

    Onde:
        cnpj : Deve ser informado um valor no padrão CNPJ com
            14 caracteres númericos, podendo conter
            sinais separadores comuns aos CNPJ\'s.

    Exemplos: 
        valida_cnpj('58.232.954/0001-93');
        valida_cnpj('75823295000193');

    ]];
    print(__usage);
end

local function valida_caracteres(__cnpj)
    if __cnpj:match("%a+") ~= nil then
        valida_cnpj.CODIGO_SAIDA = ERRO_CARACTER;
        return utils.STATE_FAIL;
    end
    return utils.STATE_SUCCESS;
end

local function verifica_tamanho(__cnpj)
    if #__cnpj ~= CNPJ_TAMANHO then
        valida_cnpj.CODIGO_SAIDA = ERRO_TAMANHO;
        return utils.STATE_FAIL;
    end
    return utils.STATE_SUCCESS;
end

local function tem_repeticao(__cnpj)
    if __cnpj == __cnpj:sub(1,1):rep(CNPJ_TAMANHO) then
        valida_cnpj.CODIGO_SAIDA = ERRO_REPETICAO;
        return utils.STATE_SUCCESS;
    end
    return utils.STATE_FAIL;
end

local function valida_digito(__cnpj)
    local __base;
    local __digito_calculado;
    local __digito_verificador;
    for __i=12, 13 do
        __base = __cnpj:sub(0,__i);
        __digito_verificador = tonumber(__cnpj:sub(__i+1,__i+1));
        __digito_calculado = calculos.modulo11(__base, 9, 2, false);
        if __digito_verificador ~= __digito_calculado then
            valida_cnpj.CODIGO_SAIDA = ERRO_DIGITO;
            return utils.STATE_FAIL;
        end
    end
    return utils.STATE_SUCCESS;
end

function valida_cnpj.valida_cnpj(__cnpj)
    if utils.esta_vazia(__cnpj, false) then
        return utils.STATE_FAIL;
    end
    if not valida_caracteres(__cnpj) then
        return utils.STATE_FAIL;
    end    
    __cnpj = utils.remove_caracter(__cnpj);
    if not verifica_tamanho(__cnpj) then
        return utils.STATE_FAIL;
    end
    if tem_repeticao(__cnpj) then
        return utils.STATE_FAIL;
    end
    if not valida_digito(__cnpj) then
        return utils.STATE_FAIL;
    end
    return utils.STATE_SUCCESS;
end

function valida_cnpj.formata(__cnpj)
    a,b,c,d,e = string.match(utils.remove_caracter(__cnpj), "(%d%d)(%d%d%d)(%d%d%d)(%d%d%d%d)(%d%d)");
    print(a.."."..b.."."..c.."/"..d.."-"..e);
end

return valida_cnpj;

File: php/Documentos/valida_cnpj.php
#!/usr/bin/env php

<?php

require_once "../Utils/calculos.php";

require_once "../Utils/utils.php";

class ValidaCnpj {

	private $__utis = new Utils();
	
	private $__calc = new Calculos();
	
	private $__CNPJ_TAMANHO = 14;

	private function print_usage() {
    	echo <<< EOF

    Sintaxe: cnpj.valida(<cnpj>)

    Onde:
        cnpj : Deve ser informado um valor no padrão cnpj com
              14 caracteres númericos, podendo conter
              sinais separadores comuns aos cnpj''s.

    Exemplos:
        valida_cnpj("12.758.232/0004-54")
        valida_cnpj("32598255000298")

EOF;
	}
    
	private function valida_caracter($__cnpj) {
		return (preg_match("/[^\d.\/-]/", $__cnpj) == 0);
	}

	private function valida_tamanho($__cnpj) {
		return ( strlen($__cnpj) == $__CNPJ_TAMANHO );
	}

	private function tem_repeticao($__cnpj) {
		$__padrao = "/[".substr($__cnpj,0,1)."]{".$__CNPJ_TAMANHO."}/";
		return ( preg_match($__padrao, $__cnpj) == 1);
	}

	private function valida_digitos($__cnpj) {
		foreach (array(12,13) as $__i) {
			$__base = substr($__cnpj, 0, $__i);
			$__dc = $__calc->modulo11($__base, 9, 2, FALSE);
			$__dv = substr($__base, $__i, 1);
			if ($__dv != $__dc) {
				return $STATE_FAIL;
			}
		}
		return $STATE_SUCCESS;
	}

	public function valida($__cnpj) {
		if ($__utils->esta_vazia($__cnpj, FALSE)) {
			return $STATE_FAIL;
		}
		if (! valida_caracter($__cnpj)) {
			return $STATE_FAIL;
		}
		$__cnpj = $__utils->remove_caracter($__cnpj);
		if (! valida_tamanho($__cnpj)) {
			return $STATE_FAIL;
		}
		if (tem_repeticao($__cnpj)) {
			return $STATE_FAIL;
		}
		if (valida_digitos($__cpf)) {
			return $STATE_FAIL;
		}
		return $STATE_SUCCESS;
	}

}

?>
